### PR TITLE
dts: add support for a bus to have multiple types

### LINKED ISSUE #29986: Add support for a single node having multiple bus types
**Is your enhancement proposal related to a problem? Please describe.**
I have `i3c-controller.yaml` that was made internally which has `bus: i3c`, and I have `i3c-devices.yaml` that has `on-bus: i3c`. But for the way i3c works, i2c devices can also compatible on the same bus (example: the i2c eeprom which includes `i2c-device.yaml` which has `on-bus: i2c` and won't look for the i3c buses. If I try to include `i2c-controller.yaml` from `i3c-controller.yaml` it complains with the error message `'bus' from included file overwritten ('i2c' replaced with 'i3c')`

**Describe the solution you'd like**
I would like to see that the `bus: ` key can be a list, and they can also be merged rather than being overwritten through `include: `. For example: `bus: [i3c, i2c]`
Some devices also would support both i3c and i2c protocols. This could pose an issue when the scripts select the proper binding. I would imagine someone would always prefer i3c over i2c bindings and the priority could be included in the order of the list where it will first try to find a binding for i3c and select that. If not, it will walk down the list and select a binding for i2c.
Also the define generated should always match the binded bus, so that when selected with `DT_ON_BUS`, it will end with what protocol it is. For example, if there is an at24 eeprom (an i2c device) on an i3c bus, it will still return i2c with `DT_ON_BUS`.

### PR DESCRIPTION
**Pull request by XenuIsWatching**
Some upcoming buses, such as i3c, have backward compatibility with i2c.
Currently, the `bus` parameter only supports a single string. This will
be problematic for binding on an i3c bus which can utilize all the
existing i2c drivers and have their `on-bus: i2c` which won't bind to
something defined as `bus: i3c`.

`bus:` would have to be defined as a list (ex. `bus: [i3c, i2c]`) for
it to bind both i3c and i2c devices. This also leaves support for it to
just be defined as a string. The preference is also encoded in the list
order so if there are bindings for both i3c and i2c, it will choose i3c
as that is first in the list.

fixes #29986

### CODE REVIEW

File: scripts/dts/edtlib.py [hunk 1]
```diff
@@ -698,7 +698,7 @@ def bus(self):
     def on_bus(self):
         "See the class docstring"
         bus_node = self.bus_node
-        return bus_node.bus if bus_node else None
+        return bus_node.bus if bus_node else bus_node
```
Comment (XenuIsWatching) [inline scripts/dts/edtlib.py:701]: This was due to an unfortunate issue with pylint. For some reason (probably due to my lack of understanding), pylint doesn't understand the type of bus_node.bus (which is to be either NoneType, str, or list). But because of the None in there, it only knows that it can be NoneType. This causing it to trip up with it sees the iteration I do after verifying that it is a list giving the error that it is `E1133: Non-iterable value on_bus is used in an iterating context (not-an-iterable)`. Wasn't quite sure how else to work around this...

**Review by galak — CHANGES_REQUESTED** [review]
You need to update the docs in dts/binding-template.yaml

File: scripts/dts/edtlib.py [hunk 2]
```diff
@@ -773,7 +773,17 @@ def _init_binding(self):
                 # specified bus (if any) and then against any bus. This is so
                 # that matching against bindings which do not specify a bus
                 # works the same way in Zephyr as it does elsewhere.
-                if (compat, on_bus) in self.edt._compat2binding:
+                if isinstance(on_bus, list) \
+                    and ((compat, elem) in self.edt._compat2binding for elem in on_bus):
```
Comment (mbolivar-nordic) [inline scripts/dts/edtlib.py:777]: Does this work?

It looks like the predicate is `isinstance(...) and (generator_expression)`. Since a generator expression is always truthy, this test doesn't look right.

Did you mean to use `any((compat, elem) in ...)` or something like that?

Would be good to get some test cases where there is and is not a matching compat,elem pair in self.edt._compat2binding just to make sure and avoid regressions.

File: scripts/dts/edtlib.py [hunk 3]
```diff
@@ -773,7 +773,17 @@ def _init_binding(self):
...
+                    for elem in on_bus:
+                        if (compat, elem) in self.edt._compat2binding:
+                            # When matching, the preference is encoded in to
+                            # bus list order, and it is to match the first one
+                            # available
+                            binding = self.edt._compat2binding[compat, elem]
+                            break
+                elif isinstance(on_bus, str) \
```
Comment (mbolivar-nordic) [inline scripts/dts/edtlib.py:785]: I think the isinstance test can be dropped here, no?

File: scripts/dts/edtlib.py [hunk 4]
```diff
@@ -1683,9 +1693,11 @@ def _check(self, require_compatible, require_description):
 
         for bus_key in "bus", "on-bus":
```
Comment (mbolivar-nordic) [inline scripts/dts/edtlib.py:1694]: Do you really need `on-bus` to be able to be a list of strings too? The commit log only mentions `bus`, and the test cases don't show how this should work for `on-bus`.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: scripts/dts/edtlib.py (hunk 3 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py:785]: It can't...

It's possible that the above if statement could fail even if it is list and it doesn't find it in the self.edt._compat2binding.

It will then attempt to do `(compat, on_bus) in self.edt._compat2binding` and will fail due to `on_bus` being a list.

File: scripts/dts/edtlib.py (hunk 4 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py:1694]: I didn't intend for `on-bus` to become a list of strings, just `bus`. I just can't think of a use case where a device would need to have multiple `on-bus`.

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py:777]: yes, you are correct. I will update it

**Comment by XenuIsWatching** [comment]
> You need to update the docs in dts/binding-template.yaml

will do 👍

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py]: In order for me to add test cases, I would have to add extra *-bus.yaml files. It doesn't look like this was being tested before 😕

File: scripts/dts/edtlib.py (hunk 4 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:1694]: > I didn't intend for `on-bus` to become a list of strings, just `bus`. I just can't think of a use case where a device would need to have multiple `on-bus`.

OK, so you will update this then, right?

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py]: >  I would have to add extra *-bus.yaml files

The current test cases were bulk ported from a pre-pytest suite that was hand-rolled.

If you want to create your own bindings directories and do special-purpose tests with DTS files that are specific to just testing this new feature to make your life easier, that would be fine with me.

File: scripts/dts/edtlib.py (hunk 4 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py]: Ah, yes... sorry, I didn't understand what you were trying to tell me earlier. Unfortunately, GitHub didn't show the full context of the line you marked so I got confused 😖 . I've split for loop up into 2 if statements now. One for "on-bus", and another for "bus".

Comment (XenuIsWatching) [reply scripts/dts/edtlib.py]: @mbolivar Is this resolved?

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (XenuIsWatching) [reply scripts/dts/edtlib.py]: @mbolivar A test case was added to test_bus(), testing the priority in the list

File: scripts/dts/edtlib.py (hunk 1 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:708]: Sorry for the delay. I needed some time to think about this and I also had a vacation.

pylint's error is confusing but there is a real problem here. People expect this to be a string or None. If you grep around the codebase you will see that making bus return a list or a string or None will break a lot of code, including gen_defines.py:

https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/gen_defines.py#L326

and gen_devicetree_rest.py:

https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/scripts/gen_devicetree_rest.py#L368

TL;DR this approach isn't workable in my opinion. Have you tested it against the internal I3C drivers you're working on? It seems like gen_defines.py in particular should crash.

I think we need a more significant API change than this, something like changing `bus` to `buses` and `on_bus` to `on_buses` in both Node and Binding. Both of these should return lists of strings, which may be empty. Then we need to update the rest of the tree to handle this.

I'll see if I can cook up a draft as part of the general ongoing conversion of this codebase to a third party library.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Comment by XenuIsWatching** [comment]
@mbolivar Sorry, I can figure out to directly reply to the earlier comment.

>It seems like gen_defines.py in particular should crash.

I have two different repos on my system. One that is synced with the upstream repo and another one that is held back a couple of versions due to dependencies. When bringing over my change I forgot the `gen_defines.py` file (there was also the `gen_legacy_defines.py` file as well, but that is no longer in the latest version), and I as was making changes to support the latest Zephyr kernel, that file got forgotten 🤦‍♂️.  I just did a quick fix there where it just checks to see if it's a list in the `str2ident` function and joins them together with an underscore. This is pushed up now...

>and gen_devicetree_rest.py:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/scripts/gen_devicetree_rest.py#L368

I'm not familiar with `gen_devicetree_rest.py` as it appears this is newer and not in the v2.3 that I've been working in. Is there a way this can be tested with the `test.dts`?

> I think we need a more significant API change than this, something like changing bus to buses and on_bus to on_buses in both Node and Binding. Both of these should return lists of strings, which may be empty. Then we need to update the rest of the tree to handle this.

If I understand what you are saying here correctly... It does feel dirty to have an item have two different types, list and string. It could make more sense to have it just only be a list that could possibly just have one item or be None.

**Comment by mbolivar-nordic** [comment]
> m not familiar with `gen_devicetree_rest.py` as it appears this is newer and not in the v2.3 that I've been working in. Is there a way this can be tested with the `test.dts`?

No, it is what is used to generate the input .rst code for the [bindings index](https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html). You can test it by building the documentation. CI for your PR is not building the documentation because the [filters](https://github.com/zephyrproject-rtos/zephyr/blob/master/.github/workflows/doc-build.yml#L8) aren't configured properly. I sent a [fix](https://github.com/zephyrproject-rtos/zephyr/pull/33885).

> Where can I find more information about this?

It's mentioned in @galak 's part of [this presentation](https://www.youtube.com/watch?v=N1NUbjVHb_U); see also https://github.com/zephyrproject-rtos/zephyr/pull/33746.

**Comment by mbolivar-nordic** [comment]
I didn't respond to this at first:

> If I understand what you are saying here correctly... It does feel dirty to have an item have two different types, list and string. It could make more sense to have it just only be a list that could possibly just have one item or be None.

From a [typing](https://docs.python.org/3/library/typing.html) perspective, this is not "only [...] a list". 

What I am thinking is a new property with a type signature that looks something like this:

```py
def buses(self) -> typing.List[str]:
    # ...
```

And similarly for `on_bus` -> `on_buses`. We need to keep using `typing.List` even though it is [deprecated](https://docs.python.org/3/library/typing.html#typing.List) because we support Pythons older than the bleeding edge. Or maybe a `typing.Set` would be more accurate, actually, if order doesn't matter. I believe order does matter since it can express preferences though ("treat me like I3C, and if you can't, fall back on I2C" could be expressed with a list `['i3c', 'i2c']`, but a set can't do that).

Anyway, a " list that could possibly just have one item or be None" would have a different signature:

```py
def buses(self) -> typing.Optional[List[str]]:
    # ...
```

I think this is not as good. An empty list still allows more natural use of things like `for bus in foo.buses` without extra null checking, but it's still falsy so you can say `if foo.buses:` if you need to.
